sw/airborne/subsystems/gps/gps_datalink.c

```diff
@@ -66,70 +50,67 @@ void gps_impl_init(void)
 // Parse the REMOTE_GPS_SMALL datalink packet
-void parse_gps_datalink_small(uint8_t num_sv, uint32_t pos_xyz, uint32_t speed_xy)
+void parse_gps_datalink_small(uint8_t num_sv, uint32_t pos_xyz, uint32_t speed_xyz, int16_t heading)
 {
-
   // Position in ENU coordinates
-  enu_pos.x = (int32_t)((pos_xyz >> 22) & 0x3FF); // bits 31-22 x position in cm
-  if (enu_pos.x & 0x200) {
-    enu_pos.x |= 0xFFFFFC00;  // fix for twos complements
+  enu_pos.x = (int32_t)((pos_xyz >> 21) & 0x7FF); // bits 31-21 x position in cm
+  if (enu_pos.x & 0x400) {
+    enu_pos.x |= 0xFFFFF800;  // sign extend for twos complements
   }
-  enu_pos.y = (int32_t)((pos_xyz >> 12) & 0x3FF); // bits 21-12 y position in cm
-  if (enu_pos.y & 0x200) {
-    enu_pos.y |= 0xFFFFFC00;  // fix for twos complements
+  enu_pos.y = (int32_t)((pos_xyz >> 10) & 0x7FF); // bits 20-10 y position in cm
+  if (enu_pos.y & 0x400) {
+    enu_pos.y |= 0xFFFFF800;  // sign extend for twos complements
   }
-  enu_pos.z = (int32_t)(pos_xyz >> 2 & 0x3FF); // bits 11-2 z position in cm
-  // bits 1 and 0 are free
-
-  // printf("ENU Pos: %u (%d, %d, %d)\n", pos_xyz, enu_pos.x, enu_pos.y, enu_pos.z);
+  enu_pos.z = (int32_t)(pos_xyz & 0x3FF); // bits 9-0 z position in cm
 
   // Convert the ENU coordinates to ECEF
-  ecef_of_enu_point_i(&ecef_pos, &tracking_ltp, &enu_pos);
-  gps.ecef_pos = ecef_pos;
+  ecef_of_enu_point_i(&gps.ecef_pos, &ltp_def, &enu_pos);
+  SetBit(gps.valid_fields, GPS_VALID_POS_ECEF_BIT);
 
-  lla_of_ecef_i(&lla_pos, &ecef_pos);
-  gps.lla_pos = lla_pos;
+  lla_of_ecef_i(&gps.lla_pos, &gps.ecef_pos);
   SetBit(gps.valid_fields, GPS_VALID_POS_LLA_BIT);
 
-  enu_speed.x = (int32_t)((speed_xy >> 22) & 0x3FF); // bits 31-22 speed x in cm/s
-  if (enu_speed.x & 0x200) {
-    enu_speed.x |= 0xFFFFFC00;  // fix for twos complements
+  enu_speed.x = (int32_t)((speed_xyz >> 21) & 0x7FF); // bits 31-21 speed x in cm/s
+  if (enu_speed.x & 0x400) {
+    enu_speed.x |= 0xFFFFF800;  // sign extend for twos complements
   }
-  enu_speed.y = (int32_t)((speed_xy >> 12) & 0x3FF); // bits 21-12 speed y in cm/s
-  if (enu_speed.y & 0x200) {
-    enu_speed.y |= 0xFFFFFC00;  // fix for twos complements
+  enu_speed.y = (int32_t)((speed_xyz >> 10) & 0x7FF); // bits 20-10 speed y in cm/s
+  if (enu_speed.y & 0x400) {
+    enu_speed.y |= 0xFFFFF800;  // sign extend for twos complements
+  }
+  enu_speed.z = (int32_t)((speed_xyz) & 0x3FF); // bits 9-0 speed z in cm/s
+  if (enu_speed.z & 0x200) {
+    enu_speed.z |= 0xFFFFFC00;  // sign extend for twos complements
   }
-  enu_speed.z = 0;
-
-  // printf("ENU Speed: %u (%d, %d, %d)\n", speed_xy, enu_speed.x, enu_speed.y, enu_speed.z);
 
-  ecef_of_enu_vect_i(&gps.ecef_vel , &tracking_ltp , &enu_speed);
+  ecef_of_enu_vect_i(&gps.ecef_vel , &ltp_def , &enu_speed);
   SetBit(gps.valid_fields, GPS_VALID_VEL_ECEF_BIT);
 
-  gps.hmsl = tracking_ltp.hmsl + enu_pos.z * 10; // TODO: try to compensate for the loss in accuracy
+  gps.ned_vel.x = enu_speed.x;
```

Comment:
I guess this should be `gps.ned_vel.x = enu_speed.y;`?
NED and ENU have x and y switched...